src/Crdt/SyncHelper.cs

```diff
@@ -0,0 +1,77 @@
+    /// <param name="remoteModel"></param>
+    /// <param name="serializerOptions"></param>
+    internal static async Task<SyncResults> SyncWith(ISyncable localModel,
+        ISyncable remoteModel,
+        JsonSerializerOptions serializerOptions)
+    {
+        if (!await localModel.ShouldSync() || !await remoteModel.ShouldSync()) return new SyncResults([], [], false);
+        var localSyncState = await localModel.GetSyncState();
+
+        var (missingFromLocal, remoteSyncState) = await remoteModel.GetChanges(localSyncState);
+        //todo abort if local and remote heads are the same
+        var (missingFromRemote, _) = await localModel.GetChanges(remoteSyncState);
+        if (localModel is DataModel && remoteModel is DataModel)
+        {
+            //cloning just to simulate the objects going over the wire
+            missingFromLocal = Clone(missingFromLocal, serializerOptions);
+            missingFromRemote = Clone(missingFromRemote, serializerOptions);
+        }
+
+        if (missingFromLocal.Length > 0)
+            await localModel.AddRangeFromSync(missingFromLocal);
+        if (missingFromRemote.Length > 0)
+            await remoteModel.AddRangeFromSync(missingFromRemote);
+        return new SyncResults(missingFromLocal, missingFromRemote, true);
+    }
+
+    internal static async Task SyncMany(ISyncable localModel, ISyncable[] remotes, JsonSerializerOptions serializerOptions)
+    {
+        var localSyncState = await localModel.GetSyncState();
+        var remoteSyncStates = new SyncState[remotes.Length];
+        for (var i = 0; i < remotes.Length; i++)
+        {
+            var remote = remotes[i];
+            var (missingFromLocal, remoteSyncState) = await remote.GetChanges(localSyncState);
+            if (localModel is DataModel && remote is DataModel)
+            {
+                //cloning just to simulate the objects going over the wire
+                missingFromLocal = Clone(missingFromLocal, serializerOptions);
+            }
+            remoteSyncStates[i] = remoteSyncState;
+            await localModel.AddRangeFromSync(missingFromLocal);
+        }
+
+        for (var i = 0; i < remotes.Length; i++)
+        {
+            var remote = remotes[i];
+            var remoteSyncState = remoteSyncStates[i];
+            var (missingFromRemote, _) = await localModel.GetChanges(remoteSyncState);
+            if (localModel is DataModel && remote is DataModel)
+            {
+                //cloning just to simulate the objects going over the wire
+                missingFromRemote = Clone(missingFromRemote, serializerOptions);
+            }
+            await remote.AddRangeFromSync(missingFromRemote);
+        }
+    }
+
+    private static T Clone<T>(this T source, JsonSerializerOptions options)
+    {
+        ArgumentNullException.ThrowIfNull(source);
+        var json = JsonSerializer.Serialize(source, options);
+        var clone = JsonSerializer.Deserialize<T>(json,options);
```

Comment:
If we Serialize and Deserialize immediately afterwards, does it matter what options we use?

Reply:
yes because we make heavy use of polymorphic serialization, but that only works when it knows what types to use.